<a href="https://colab.research.google.com/github/AjjayK/Amex-Credit-Default-Prediction/blob/main/EDA_and_Deep_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#American Express &ndash; Default Prediction
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:-20.45pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;text-indent:41.75pt;'><span style='font-family:"Open Sans",sans-serif;'>Predict if a customer will default in the future</span></p>

---

<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:-20.45pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;text-indent:41.75pt;'><span style='font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:-20.45pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;text-indent:41.75pt;'><strong><span style='font-size:19px;line-height:107%;font-family:"Open Sans",sans-serif;'>About American Express</span></strong></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>American Express Company (Amex) is an American multinational corporation specialized in payment card services headquartered at 200 Vesey Street in the Battery Park City neighborhood of Lower Manhattan in New York City. The company was founded in 1850 and is one of the 30 components of the Dow Jones Industrial Average. The company&apos;s logo, adopted in 1958, is a gladiator or centurion whose image appears on the company&apos;s well-known traveler&apos;s cheques, charge cards, and credit cards.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>During the 1980s, Amex invested in the brokerage industry, acquiring what became, in increments, Shearson Lehman Hutton and then divesting these into what became Smith Barney Shearson (owned by Primerica) and a revived Lehman Brothers. By 2008 neither the Shearson nor the Lehman name existed.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>In 2016, credit cards using the American Express network accounted for 22.9% of the total dollar volume of credit card transactions in the United States. As of December 31, 2021, the company had 121.7 million cards in force, including 56.4 million cards in force in the United States, each with an average annual spending of $20,392.</span></p>

<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:0cm;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>In 2017, Forbes named American Express as the 23rd most valuable brand in the world (and the highest within financial services), estimating the brand to be worth US$24.5 billion. In 2020, Fortune magazine ranked American Express at number 9 on their Fortune List of the Top 100 Companies to Work For in 2020 based on an employee survey of satisfaction</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>

---

<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:19px;line-height:107%;font-family:"Open Sans",sans-serif;'>About Competition</span></strong></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>American Express hosted this competition in Kaggle. Quoted below is the description of the competition provided in the Kaggle site.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>Whether out at a restaurant or buying tickets to a concert, modern life counts on the convenience of a credit card to make daily purchases. It saves us from carrying large amounts of cash and also can advance a full purchase that can be paid overtime. How do card issuers know we&rsquo;ll pay back what we charge? That&rsquo;s a complex problem with many existing solutions&mdash;and even more potential improvements, to be explored in this competition.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>Credit default prediction is central to managing risk in a consumer lending business. Credit default prediction allows lenders to optimize lending decisions, which leads to a better customer experience and sound business economics. Current models exist to help manage risk. But it&apos;s possible to create better models that can outperform those currently in use.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>American Express is a globally integrated payments company. The largest payment card issuer in the world, they provide customers with access to products, insights, and experiences that enrich lives and build business success.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>In this competition, you&rsquo;ll apply your machine learning skills to predict credit default. Specifically, you will leverage an industrial scale data set to build a machine learning model that challenges the current model in production. Training, validation, and testing datasets include time-series behavioral data and anonymized customer profile information. You&apos;re free to explore any technique to create the most powerful model, from creating features to using the data in a more organic way within a model.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>If successful, you&apos;ll help create a better customer experience for cardholders by making it easier to be approved for a credit card. Top solutions could challenge the credit default prediction model used by the world&apos;s largest payment card issuer&mdash;earning you cash prizes, the opportunity to interview with American Express, and potentially a rewarding new career.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>


---


<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:19px;line-height:107%;font-family:"Open Sans",sans-serif;'>About Dataset</span></strong></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>The objective of this competition is to predict the probability that a customer does not pay back their credit card balance amount in the future based on their monthly customer profile. The target binary variable is calculated by observing 18 months performance window after the latest credit card statement, and if the customer does not pay due amount in 120 days after their latest statement date it is considered a default event.</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>The dataset contains aggregated profile features for each customer at each statement date. Features are anonymized and normalized, and fall into the following general categories:</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>D_* = Delinquency variables</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>S_* = Spend variables</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>P_* = Payment variables</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>B_* = Balance variables</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>R_* = Risk variables</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>with the following features being categorical:</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>[&apos;B_30&apos;, &apos;B_38&apos;, &apos;D_114&apos;, &apos;D_116&apos;, &apos;D_117&apos;, &apos;D_120&apos;, &apos;D_126&apos;, &apos;D_63&apos;, &apos;D_64&apos;, &apos;D_66&apos;, &apos;D_68&apos;]</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>The task is to predict, for each customer_ID, the probability of a future payment default (target = 1).</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>Note that the negative class has been subsampled for this dataset at 5%, and thus receives a 20x weighting in the scoring metric</span></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>&nbsp;</span></p>

---

<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:19px;line-height:107%;font-family:"Open Sans",sans-serif;'>About Notebook</span></strong></p>
<p style='margin-top:0cm;margin-right:0cm;margin-bottom:8.0pt;margin-left:21.3pt;line-height:107%;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;line-height:107%;font-family:"Open Sans",sans-serif;'>This notebook consists of Exploratory Data Analysis (EDA) tool for analyzing the variable characteristics and Deep Neural Decision Tree, Deep Neural Decision Forest models for predicting the credit default.</span></p>

###**Initializing the notebook**
The notebook runs on Google Colab. Google Colab uses Python 3.7.

In [ ]:
#Mouting Google Drive
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
#Importing the libraries used in this notebook
import os

import numpy as np
import pandas as pd

'''#Data Viz
import plotly
import plotly.graph_objects as go
from dash import Dash, dcc, html, Input, Output
from plotly.subplots import make_subplots
import dash_bootstrap_components as dbc'''

#Tensorflow
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import math

'''#LGBM
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedKFold 
from sklearn.metrics import roc_auc_score, roc_curve, auc
from lightgbm import LGBMClassifier, early_stopping, log_evaluation'''

#System
import warnings, gc
warnings.filterwarnings("ignore")

###**Loading the Dataset**
Dataset was downloaded from Kaggle and is copied to the drive

In [ ]:
!pip install pyarrow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#Loading the training data
train_data = pd.read_feather("/content/drive/MyDrive/Amex Kaggle/train_data.ftr")
train_data = train_data.groupby('customer_ID').tail(1).set_index('customer_ID')

###**EDA Tool**

In [ ]:
#Column stratification
col = train_data.columns.to_list()
cat_features = ['B_30', 'B_38', 'D_114', 'D_116', 'D_117', 'D_120', 'D_126', 'D_63', 'D_64', 'D_66', 'D_68']
non_cat_features = [x for x in col if x not in cat_features]

delinquency_variables = [i for i in col if i.startswith('D_')]
spend_variables = [i for i in col if (i.startswith('S_') and i != 'S_2')]
payment_variables = [i for i in col if i.startswith('P_')]
balance_variables = [i for i in col if i.startswith('B_')]
risk_variables = [i for i in col if i.startswith('R_')]

In [ ]:
#Creating a html site using Dash Library for EDA
external_stylesheets = [dbc.themes.SOLAR]
app = Dash(__name__, external_stylesheets=external_stylesheets)
all_options = {'delinquency_variables':delinquency_variables,'spend_variables':spend_variables,'payment_variables':payment_variables,
              'balance_variables':balance_variables,'risk_variables':risk_variables}


app.layout = html.Div([
    html.H3("Variable Information Dashboard"),
    html.Div([
    html.H6("Select the variable type"),
    dcc.Dropdown(
    list(all_options.keys()),
    'delinquency_variables',
    id = 'variable-type'),
    html.H6("Select the variable"),
    dcc.Dropdown(id = 'variable')]),
    
    dcc.Graph(id = 'dist-plot'),
    dcc.Graph(id = 'heatmaps')
])

@app.callback(
Output('variable','options'),
Input('variable-type','value'))
def set_variable_options(variable_type):
    return [{'label': i, 'value': i} for i in all_options[variable_type]]

@app.callback(
Output('variable','value'),
    Input('variable','options'))
def set_variable_value(available_options):
    return available_options[0]['value']

@app.callback(
Output('dist-plot','figure'),
Output('heatmaps','figure'),
Input('variable-type','value'),
Input('variable','value'))
def hist_plot(variable_type, variable):
    fig1 = make_subplots(rows=1, cols=2,column_widths=[0.7, 0.3],
    subplot_titles=("Histogram for "+ variable_type +" "+variable,"Boxplot for "+ variable_type +" "+variable))

    traces_nd = go.Histogram(name= "Non-Defaulting", x = train_data[variable][train_data['target']==0].to_list(),histnorm='percent',nbinsx = 50, marker_color = 'indianred' )
    traces_d = go.Histogram(name = "Defaulting", x  = train_data[variable][train_data['target']==1].to_list(),histnorm='percent',nbinsx = 50, marker_color = 'lightseagreen' )
    fig1.append_trace(traces_nd,1,1)
    fig1.append_trace(traces_d,1,1)

    trace_box_nd = go.Box(y = train_data[variable][train_data['target']==0].to_list(), name = "Non-Defaulting", marker_color = 'indianred')
    trace_box_d = go.Box(y = train_data[variable][train_data['target']==1].to_list(), name = "Defaulting", marker_color = 'lightseagreen')


    fig1.append_trace(trace_box_nd,1,2)
    fig1.append_trace(trace_box_d,1,2)

    fig1.update_layout(barmode='overlay')
    fig1.update_traces(opacity=0.75)
    
    cor = train_data[all_options[variable_type]].corr()
    cor = cor[variable].sort_values(ascending = False)

    fig2 = go.Figure()
    fig2.add_trace(go.Bar(x = cor[cor>=0], y = cor.index[cor>=0], orientation = 'h', marker_color='rgb(167, 255, 235)', marker_line_color='rgb(0, 105, 92)',
                      marker_line_width=1.5, opacity=0.6, showlegend = False, text = cor[cor>=0], textposition = 'outside', texttemplate = '%{x:.3f}' ))
    fig2.add_trace(go.Bar(x=cor[cor<0].sort_values(ascending = False), y = cor[cor<0].sort_values(ascending = False).index,orientation = 'h',marker_color='rgb(255, 138, 128)', marker_line_color='rgb(198, 40, 40)',
                      marker_line_width=1.5, opacity=0.6, showlegend = False, text = cor[cor<0].sort_values(ascending = False), textposition = 'outside',texttemplate = '%{x:.3f}'))
    fig2.update_layout(title="Correlation Plot for " + variable_type +" "+variable,
                      xaxis_title="Correlation", margin=dict(l=150),
                      height=2500)

    
    return fig1, fig2

NameError: ignored

In [ ]:
#Firing the Dash app
if __name__ == '__main__':
    app.run_server(debug=False)


#Click on the hyperlink to see the visualization

###**Preparing the dataset for deep learning and machine learning models**

In [ ]:
#Dropping the customer ID column
train_data.drop(columns = ['S_2'], inplace = True)
train_data.to_csv('train_data_csv.csv', index = False, header = False)


test_data = pd.read_feather("/content/drive/MyDrive/Amex Kaggle/test_data.ftr")
test_data = test_df.groupby('customer_ID').tail(1).set_index('customer_ID')
test_data['target'] = np.nan
test_data.drop(columns = ['S_2'], inplace = True)
test_data.to_csv('test_data_csv.csv', index = False, header = False)

###**Classification with Neural Decision Tree and Forest**

In [ ]:
#Segregating categorical and numerical features
CSV_HEADER = train_data.columns.to_list()
cat_features = ['B_30', 'B_38', 'D_114', 'D_116', 'D_117', 'D_120', 'D_126', 'D_63', 'D_64', 'D_66', 'D_68']
NUMERIC_FEATURE_NAMES = [x for x in CSV_HEADER if x not in cat_features]
NUMERIC_FEATURE_NAMES.remove('target')

In [ ]:
# A list of the numerical feature names.

# A dictionary of the categorical features and their vocabulary.

CATEGORICAL_FEATURES_WITH_VOCABULARY = {
    "B_30": sorted(list(train_data["B_30"].unique())),
    "B_38": sorted(list(train_data["B_38"].unique())),
    "D_114": sorted(list(train_data["D_114"].unique())),
    "D_116": sorted(list(train_data["D_116"].unique())),
    "D_117": sorted(list(train_data["D_117"].unique())),
    "D_120": sorted(list(train_data["D_120"].unique())),
    "D_126": sorted(list(train_data["D_126"].unique())),
    "D_63": sorted(list(train_data["D_63"].unique())),
    "D_64": sorted(list(train_data["D_64"].unique())),
    "D_66": sorted(list(train_data["D_66"].unique())),
    "D_68": sorted(list(train_data["D_68"].unique()))
}
# A list of the columns to ignore from the dataset.
IGNORE_COLUMN_NAMES = ["customer_ID"]
# A list of the categorical feature names.
CATEGORICAL_FEATURE_NAMES = list(CATEGORICAL_FEATURES_WITH_VOCABULARY.keys())
# A list of all the input features.
FEATURE_NAMES = NUMERIC_FEATURE_NAMES + CATEGORICAL_FEATURE_NAMES
# A list of column default values for each feature.
COLUMN_DEFAULTS = [
    [0.0] if feature_name in NUMERIC_FEATURE_NAMES + IGNORE_COLUMN_NAMES else ["NA"]
    for feature_name in CSV_HEADER
]
# The name of the target feature.
TARGET_FEATURE_NAME = "target"
# A list of the labels of the target features.
TARGET_LABELS = ["0", "1"]

####**Create tf.data.Dataset objects for training and validation**
We create an input function to read and parse the file, and convert features and labels into a [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) for training and validation. We also preprocess the input by mapping the target label to an index.

In [ ]:
from tensorflow.keras.layers import StringLookup

target_label_lookup = StringLookup(
    vocabulary=TARGET_LABELS, mask_token=None, num_oov_indices=0
)
target_label_lookup2 = StringLookup(
    vocabulary=TARGET_LABELS, mask_token=None, num_oov_indices=1
)

def get_dataset_from_csv(csv_file_path, shuffle=False, batch_size=128):
    dataset = tf.data.experimental.make_csv_dataset(
        csv_file_path,
        batch_size=batch_size,
        column_defaults=COLUMN_DEFAULTS,
        label_name=TARGET_FEATURE_NAME,
        num_epochs=1,
        na_value="?",
        shuffle=shuffle,
        header = False,
        column_names = CSV_HEADER
    ).map(lambda features, target: (features, target_label_lookup(target)))
    return dataset.cache()

def get_dataset_from_csv2(csv_file_path, shuffle=False, batch_size=128):
    dataset = tf.data.experimental.make_csv_dataset(
        csv_file_path,
        batch_size=batch_size,
        column_defaults=COLUMN_DEFAULTS,
        label_name=TARGET_FEATURE_NAME,
        num_epochs=1,
        na_value="?",
        shuffle=shuffle,
        header = False,
        column_names = CSV_HEADER
    ).map(lambda features, target: (features, target_label_lookup2(target)))
    return dataset.cache()

####**Create model inputs**

In [ ]:
def create_model_inputs():
    inputs = {}
    for feature_name in FEATURE_NAMES:
        if feature_name in NUMERIC_FEATURE_NAMES:
            inputs[feature_name] = layers.Input(
                name=feature_name, shape=(), dtype=tf.float32
            )
        else:
            inputs[feature_name] = layers.Input(
                name=feature_name, shape=(), dtype=tf.string
            )
    return inputs

####**Encode input features**

In [ ]:
def encode_inputs(inputs):
    encoded_features = []
    for feature_name in inputs:
        if feature_name in CATEGORICAL_FEATURE_NAMES:
            vocabulary = [str(x) for x in CATEGORICAL_FEATURES_WITH_VOCABULARY[feature_name]]
            # Create a lookup to convert a string values to an integer indices.
            # Since we are not using a mask token, nor expecting any out of vocabulary
            # (oov) token, we set mask_token to None and num_oov_indices to 0.
            lookup = StringLookup(
                vocabulary=vocabulary, mask_token=None, num_oov_indices=1
            )
            # Convert the string input values into integer indices.
            value_index = lookup(inputs[feature_name])
            embedding_dims = int(math.sqrt(lookup.vocabulary_size()))
            # Create an embedding layer with the specified dimensions.
            embedding = layers.Embedding(
                input_dim=lookup.vocabulary_size(), output_dim=embedding_dims
            )
            # Convert the index values to embedding representations.
            encoded_feature = embedding(value_index)
        else:
            # Use the numerical features as-is.
            encoded_feature = inputs[feature_name]
            if inputs[feature_name].shape[-1] is None:
                encoded_feature = tf.expand_dims(encoded_feature, -1)

        encoded_features.append(encoded_feature)

    encoded_features = layers.concatenate(encoded_features)
    return encoded_features

####**Deep Neural Decision Tree**
A neural decision tree model has two sets of weights to learn. The first set is pi, which represents the probability distribution of the classes in the tree leaves. The second set is the weights of the routing layer decision_fn, which represents the probability of going to each leave. The forward pass of the model works as follows:



1.   The model expects input features as a single vector encoding all the features of an instance in the batch. This vector can be generated from a Convolution Neural Network (CNN) applied to images or dense transformations applied to structured data features.
2.   The model first applies a used_features_mask to randomly select a subset of input features to use.
3.   Then, the model computes the probabilities (mu) for the input instances to reach the tree leaves by iteratively performing a stochastic routing throughout the tree levels.
4.   Finally, the probabilities of reaching the leaves are combined by the class probabilities at the leaves to produce the final outputs.

In [ ]:
class NeuralDecisionTree(keras.Model):
    def __init__(self, depth, num_features, used_features_rate, num_classes):
        super(NeuralDecisionTree, self).__init__()
        self.depth = depth
        self.num_leaves = 2 ** depth
        self.num_classes = num_classes

        # Create a mask for the randomly selected features.
        num_used_features = int(num_features * used_features_rate)
        one_hot = np.eye(num_features)
        sampled_feature_indicies = np.random.choice(
            np.arange(num_features), num_used_features, replace=False
        )
        self.used_features_mask = one_hot[sampled_feature_indicies]

        # Initialize the weights of the classes in leaves.
        self.pi = tf.Variable(
            initial_value=tf.random_normal_initializer()(
                shape=[self.num_leaves, self.num_classes]
            ),
            dtype="float32",
            trainable=True,
        )

        # Initialize the stochastic routing layer.
        self.decision_fn = layers.Dense(
            units=self.num_leaves, activation="sigmoid", name="decision"
        )

    def call(self, features):
        batch_size = tf.shape(features)[0]

        # Apply the feature mask to the input features.
        features = tf.matmul(
            features, self.used_features_mask, transpose_b=True
        )  # [batch_size, num_used_features]
        # Compute the routing probabilities.
        decisions = tf.expand_dims(
            self.decision_fn(features), axis=2
        )  # [batch_size, num_leaves, 1]
        # Concatenate the routing probabilities with their complements.
        decisions = layers.concatenate(
            [decisions, 1 - decisions], axis=2
        )  # [batch_size, num_leaves, 2]

        mu = tf.ones([batch_size, 1, 1])

        begin_idx = 1
        end_idx = 2
        # Traverse the tree in breadth-first order.
        for level in range(self.depth):
            mu = tf.reshape(mu, [batch_size, -1, 1])  # [batch_size, 2 ** level, 1]
            mu = tf.tile(mu, (1, 1, 2))  # [batch_size, 2 ** level, 2]
            level_decisions = decisions[
                :, begin_idx:end_idx, :
            ]  # [batch_size, 2 ** level, 2]
            mu = mu * level_decisions  # [batch_size, 2**level, 2]
            begin_idx = end_idx
            end_idx = begin_idx + 2 ** (level + 1)

        mu = tf.reshape(mu, [batch_size, self.num_leaves])  # [batch_size, num_leaves]
        probabilities = keras.activations.softmax(self.pi)  # [num_leaves, num_classes]
        outputs = tf.matmul(mu, probabilities)  # [batch_size, num_classes]
        return outputs

In [93]:
#Fitting the Neural Decision Tree
learning_rate = 0.01
batch_size = 1000
num_epochs = 5
hidden_units = [64, 64]


def run_experiment(model):

    model.compile(
        optimizer=keras.optimizers.Adam(learning_rate=learning_rate),
        loss=keras.losses.SparseCategoricalCrossentropy(),
        metrics=[keras.metrics.SparseCategoricalAccuracy()],
    )

    print("Start training the model...")
    train_dataset = get_dataset_from_csv(
        'train_data_csv.csv', shuffle=False, batch_size=batch_size
    )

    model.fit(train_dataset, epochs=num_epochs)
    print("Model training finished")

    return model

In [94]:
#Creating the model
num_trees = 10
depth = 8
used_features_rate = 1.0
num_classes = len(TARGET_LABELS)


def create_tree_model():
    inputs = create_model_inputs()
    features = encode_inputs(inputs)
    features = layers.BatchNormalization()(features)
    num_features = features.shape[1]

    tree = NeuralDecisionTree(depth, num_features, used_features_rate, num_classes)

    outputs = tree(features)
    model = keras.Model(inputs=inputs, outputs=outputs)
    return model


tree_model = create_tree_model()
decision_tree = run_experiment(tree_model)

Start training the model...
Epoch 1/5
459/459 [==============================] - 73s 150ms/step - loss: 0.2909 - sparse_categorical_accuracy: 0.8938
Epoch 2/5
459/459 [==============================] - 10s 22ms/step - loss: 0.2325 - sparse_categorical_accuracy: 0.8988
Epoch 3/5
459/459 [==============================] - 10s 21ms/step - loss: 0.2271 - sparse_categorical_accuracy: 0.9004
Epoch 4/5
459/459 [==============================] - 10s 21ms/step - loss: 0.2237 - sparse_categorical_accuracy: 0.9021
Epoch 5/5
459/459 [==============================] - 10s 21ms/step - loss: 0.2211 - sparse_categorical_accuracy: 0.9038
Model training finished


In [95]:
#Predicting the response using model for training data and test data
decision_tree_ypred_train = decision_tree.predict(get_dataset_from_csv(
        'train_data_csv.csv', shuffle=False, batch_size=batch_size
    ))

decision_tree_ypred_test = decision_tree.predict(get_dataset_from_csv2(
        'test_data_csv.csv', shuffle=False, batch_size=batch_size
    ))


####**Deep Neural Decision Forest**
The neural decision forest model consists of a set of neural decision trees that are trained simultaneously. The output of the forest model is the average outputs of its trees.

In [96]:
class NeuralDecisionForest(keras.Model):
    def __init__(self, num_trees, depth, num_features, used_features_rate, num_classes):
        super(NeuralDecisionForest, self).__init__()
        self.ensemble = []
        # Initialize the ensemble by adding NeuralDecisionTree instances.
        # Each tree will have its own randomly selected input features to use.
        for _ in range(num_trees):
            self.ensemble.append(
                NeuralDecisionTree(depth, num_features, used_features_rate, num_classes)
            )

    def call(self, inputs):
        # Initialize the outputs: a [batch_size, num_classes] matrix of zeros.
        batch_size = tf.shape(inputs)[0]
        outputs = tf.zeros([batch_size, num_classes])

        # Aggregate the outputs of trees in the ensemble.
        for tree in self.ensemble:
            outputs += tree(inputs)
        # Divide the outputs by the ensemble size to get the average.
        outputs /= len(self.ensemble)
        return outputs

In [106]:
#Fitting the forest model
num_trees = 25
depth = 5
used_features_rate = 1
num_classes = len(TARGET_LABELS)

def create_forest_model():
    inputs = create_model_inputs()
    features = encode_inputs(inputs)
    features = layers.BatchNormalization()(features)
    num_features = features.shape[1]

    forest_model = NeuralDecisionForest(
        num_trees, depth, num_features, used_features_rate, num_classes
    )

    outputs = forest_model(features)
    model = keras.Model(inputs=inputs, outputs=outputs)
    return model


forest_model = create_forest_model()

fitted_forest = run_experiment(forest_model)

Start training the model...
Epoch 1/5
459/459 [==============================] - 87s 156ms/step - loss: 0.2867 - sparse_categorical_accuracy: 0.8953
Epoch 2/5
459/459 [==============================] - 23s 49ms/step - loss: 0.2283 - sparse_categorical_accuracy: 0.9003
Epoch 3/5
459/459 [==============================] - 23s 49ms/step - loss: 0.2228 - sparse_categorical_accuracy: 0.9018
Epoch 4/5
459/459 [==============================] - 23s 49ms/step - loss: 0.2190 - sparse_categorical_accuracy: 0.9033
Epoch 5/5
459/459 [==============================] - 22s 49ms/step - loss: 0.2159 - sparse_categorical_accuracy: 0.9043
Model training finished


In [111]:
#Predicting the response using model for training data and test data
forest_y_pred_train = fitted_forest.predict(get_dataset_from_csv(
        'train_data_csv.csv', shuffle=False, batch_size=batch_size
    ))
forest_y_pred_test = fitted_forest.predict(get_dataset_from_csv2(
        'test_data_csv.csv', shuffle=False, batch_size=batch_size
    ))


###**Model Evaluation**
#####**Amex Metrics**
The evaluation metric, M, for this competition is the mean of two measures of rank ordering: Normalized Gini Coefficient, G, and default rate captured at 4%, D.

M=0.5⋅(G+D)
The default rate captured at 4% is the percentage of the positive labels (defaults) captured within the highest-ranked 4% of the predictions, and represents a Sensitivity/Recall statistic.

For both of the sub-metrics G and D, the negative labels are given a weight of 20 to adjust for downsampling.

This metric has a maximum value of 1.0.
#####**Sparse Categorical Accuracy**
Since interpretation of Amex metrics is difficult, sparse categorical accuracy is estimated for the training data. As 'target' values are not avilable for test data, we only evaluate this metric for training data. It was found that higher the categorical accruacy on training data, higher the Amex metrics on test data. Hence, high value of the categorical accuracy on training data should give a better estimate for higher categorical accuracy for test data.

In [99]:
#Sparse Categorical Accuracy
def sparse_categorical_accuracy(y_true, y_pred):
  acc = np.dot(1, np.equal(y_true, np.argmax(y_pred, axis=1)))
  return (sum(acc)/len(acc))*100

In [100]:
#Amex Metrics
def amex_metric(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:

    def top_four_percent_captured(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        df = (pd.concat([y_true, y_pred], axis='columns')
              .sort_values('prediction', ascending=False))
        df['weight'] = df['target'].apply(lambda x: 20 if x==0 else 1)
        four_pct_cutoff = int(0.04 * df['weight'].sum())
        df['weight_cumsum'] = df['weight'].cumsum()
        df_cutoff = df.loc[df['weight_cumsum'] <= four_pct_cutoff]
        return (df_cutoff['target'] == 1).sum() / (df['target'] == 1).sum()
        
    def weighted_gini(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        df = (pd.concat([y_true, y_pred], axis='columns')
              .sort_values('prediction', ascending=False))
        df['weight'] = df['target'].apply(lambda x: 20 if x==0 else 1)
        df['random'] = (df['weight'] / df['weight'].sum()).cumsum()
        total_pos = (df['target'] * df['weight']).sum()
        df['cum_pos_found'] = (df['target'] * df['weight']).cumsum()
        df['lorentz'] = df['cum_pos_found'] / total_pos
        df['gini'] = (df['lorentz'] - df['random']) * df['weight']
        return df['gini'].sum()

    def normalized_weighted_gini(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        y_true_pred = y_true.rename(columns={'target': 'prediction'})
        return weighted_gini(y_true, y_pred) / weighted_gini(y_true, y_true_pred)

    g = normalized_weighted_gini(y_true, y_pred)
    d = top_four_percent_captured(y_true, y_pred)

    return 0.5 * (g + d)

In [112]:
#Evaluating the sparse categorical accuracy for training data
y_true = train_data['target']
print("Sparse categorical accuracy for deep decision tree model: {:.5f}%".format(sparse_categorical_accuracy(y_true,decision_tree_ypred_train)))
print("Sparse categorical accuracy for deep forest model: {:.5f}%".format(sparse_categorical_accuracy(y_true,forest_y_pred_train)))

Sparse categorical accuracy for deep decision tree model: 90.45527%
Sparse categorical accuracy for deep forest model: 90.58667%


In [113]:
#Evaluating the amex metrics for training data
print("Amex metric for deep decision tree model: {:.5f}".format(amex_metric(y_true.to_frame().reset_index(),pd.DataFrame(np.argmax(decision_tree_ypred_train, axis=1), columns = ['prediction']))))
print("Amex metric for deep forest model: {:.5f}".format(amex_metric(y_true.to_frame().reset_index(),pd.DataFrame(np.argmax(forest_y_pred_train, axis=1), columns = ['prediction']))))

Amex metric for deep decision tree model: 0.59020
Amex metric for deep forest model: 0.59137


In [114]:
#Evaluating the amex metrics for training data
print("Amex metric for deep decision tree model: {:.5f}".format(amex_metric(y_true.to_frame(),pd.DataFrame(decision_tree_ypred_train[:,1], columns = ['prediction'], index = train_data.index))))
print("Amex metric for deep forest model: {:.5f}".format(amex_metric(y_true.to_frame(),pd.DataFrame(forest_y_pred_train[:,1], columns = ['prediction'],index = train_data.index))))

Amex metric for deep decision tree model: 0.79070
Amex metric for deep forest model: 0.80014


###Test Data Evaluation
Using Kaggle API to submit to the competition. Enter your kaggle username and kaggle key

In [115]:
decision_tree_sub = pd.DataFrame(decision_tree_ypred_test[:,1], columns = ['prediction'], index =test_df.index)
decision_tree_sub.to_csv('decision_tree_sub.csv')


In [116]:
forest_sub = pd.DataFrame(forest_y_pred_test[:,1], columns = ['prediction'], index =test_df.index)
forest_sub.to_csv('forest.csv')

In [80]:
import os
os.environ['KAGGLE_username'] = username
os.environ['KAGGLE_key'] = key
import kaggle

In [117]:
!kaggle competitions submit -c amex-default-prediction -f /content/decision_tree_sub.csv -m decision_tree_sub

100% 67.6M/67.6M [00:06<00:00, 11.7MB/s]
Successfully submitted to American Express - Default Prediction

In [118]:
!kaggle competitions submit -c amex-default-prediction -f /content/forest.csv -m forest_sub

100% 67.6M/67.6M [00:06<00:00, 11.6MB/s]
Successfully submitted to American Express - Default Prediction

In [119]:
!kaggle competitions submissions -c amex-default-prediction

fileName               date                 description                status    publicScore  privateScore  
---------------------  -------------------  -------------------------  --------  -----------  ------------  
forest.csv             2022-09-10 18:42:36  forest_sub                 complete  0.78227      0.79177       
decision_tree_sub.csv  2022-09-10 18:42:27  decision_tree_sub          complete  0.77162      0.78181       
lgbm_out.csv           2022-09-10 18:41:38  lgbm                       complete  0.78604      0.79546       
lgbm_out.csv           2022-09-10 18:20:40  Submitting again for lgbm  complete  0.78604      0.79546       
lgbm_out.csv           2022-09-10 18:17:53  lgbm baseline              complete  0.57899      0.56586       
decision_tree_sub.csv  2022-09-10 18:01:25  decision_tree_sub          complete  0.56312      0.56595       
submission2 (7).csv    2022-08-27 17:58:53                             complete  0.77662      0.78733       
submission2 (6).csv

###**Conclusion**

1.   LGBM (Model in other notebook) stands out to be better in terms of metrics (Score - 0.79546 and Categorical Accuracy on training data - 92.87%)
2.   Amex Metric for Deep Decision Forest Model was better than Deep Neural Decision Tree since Forest Model is an ensemble model

Overall results are given in the table below

<table style="margin-left:21.3pt;border-collapse:collapse;border:none;">
    <tbody>
        <tr>
            <td style="width: 134.85pt;border: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Model</span></strong></p>
            </td>
            <td style="width: 134.85pt;border-top: 1pt solid windowtext;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-image: initial;border-left: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Categorical Accuracy on Training Data</span></strong></p>
            </td>
            <td style="width: 134.9pt;border-top: 1pt solid windowtext;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-image: initial;border-left: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Amex Metric on Training Data</span></strong></p>
            </td>
            <td style="width: 134.9pt;border-top: 1pt solid windowtext;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-image: initial;border-left: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><strong><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Amex Metric on Test Data</span></strong></p>
            </td>
        </tr>
        <tr>
            <td style="width: 134.85pt;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-left: 1pt solid windowtext;border-image: initial;border-top: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>LGBM</span></p>
            </td>
            <td style="width: 134.85pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>92.8759%</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.6813</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.7954</span></p>
            </td>
        </tr>
        <tr>
            <td style="width: 134.85pt;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-left: 1pt solid windowtext;border-image: initial;border-top: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Deep Neural Decision Tree</span></p>
            </td>
            <td style="width: 134.85pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>90.4552%</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.7907</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.7818</span></p>
            </td>
        </tr>
        <tr>
            <td style="width: 134.85pt;border-right: 1pt solid windowtext;border-bottom: 1pt solid windowtext;border-left: 1pt solid windowtext;border-image: initial;border-top: none;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>Deep Neural Decision Forest</span></p>
            </td>
            <td style="width: 134.85pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>90.5866%</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.8001</span></p>
            </td>
            <td style="width: 134.9pt;border-top: none;border-left: none;border-bottom: 1pt solid windowtext;border-right: 1pt solid windowtext;padding: 0cm 5.4pt;vertical-align: top;">
                <p style='margin-top:0cm;margin-right:0cm;margin-bottom:0cm;margin-left:0cm;line-height:normal;font-size:15px;font-family:"Calibri",sans-serif;text-align:center;'><span style='font-size:16px;font-family:"Open Sans",sans-serif;'>0.7917</span></p>
            </td>
        </tr>
    </tbody>
</table>